improver/memprofile.py

```diff
@@ -0,0 +1,125 @@
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module containing maximum memory profiling utilities."""
+
+import tracemalloc as tm
+from queue import Queue, Empty
+from threading import Thread
+from resource import getrusage, RUSAGE_SELF
+from datetime import datetime
+
+
+def memory_profile_start(outfile_prefix, timeout_time=0.1):
+    """Starts the memory tracking profiler.
+
+    Args:
+        outfile_prefix (str):
+            Prefix for the generated output. 2 files will
+            be generated: *_SNAPSHOT and *_MAX_TRACKER.
+        timeout_time (float):
+            Time the queue will look for an input.
+
+    Returns:
+        Active Thread tracking the memory.
+        Active Queue for communication to the thread.
+    """
+    queue = Queue()
+    thread = Thread(target=memory_monitor, args=(queue, timeout_time,
+                                                 outfile_prefix))
+    thread.start()
+    return thread, queue
+
+
+def memory_profile_end(queue, thread):
+    """Ends the memory tracking profiler.
+
+    Args:
+        queue (queue.Queue):
+            Active queue instance to communicate with active thread.
+        thread (thread.Thread):
+            Active thread instance running memory tracking.
+    """
+    queue.put('Stop')
+    thread.join()
+
+
+def memory_monitor(queue, timeout_time, outfile_prefix):
+    """Function to track memory usage, should be run in a separate
+    thread to the main program.
+
+    Args:
+        queue (queue.Queue):
+            Active queue instance to communicate with the thread.
+        timeout_time (float):
+            Time the queue will look for in input.
+        outfile_prefix (str):
+            Prefix for the generated output. 2 files will
+            be generated: *_SNAPSHOT and *_MAX_TRACKER.
+    """
+    tm.start()
+    old_max = 0
+    snapshot = None
+
+    fout = open("{}_MAX_TRACKER".format(outfile_prefix), 'w')
+    b2mb = 1 / 1048576
+
+    while True:
+        try:
+            queue.get(timeout=timeout_time)
+            snapshot.dump("{}_SNAPSHOT".format(outfile_prefix))
+            fout.close()
+            tm.stop()
+            return
+        except Empty:
+            max_rss = getrusage(RUSAGE_SELF).ru_maxrss
```

Comment:
On Linux, [maxrss is in kilobytes](https://manpages.debian.org/buster/manpages-dev/getrusage.2.en.html) and Python passes this through in whatever units the OS reports it in. There's an accompanying open/unresolved Python bug https://bugs.python.org/issue30513 relating to this issue.

If maxrss is in kilobytes, the `max_rss * b2mb` conversion to megabytes for printing below will be incorrect. Suggest checking that the result printed in _MAX_TRACKER corresponds with what `/usr/bin/time -v` (the command generally has additional features compared to the shell `time` builtin) reports.

Reply:
well... that's annoying - you're completely right, I'll add in a sys.platform check